Add loading screen component

gcsecsey opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes https://github.com/Automattic/dotcom-forge/issues/7451

## Proposed Changes

* Add loading screen component
* Show loading screen while fetching performance metrics

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* We're updating the Site Profiler UI

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

- Apply the patch D150503-code and sandbox public-api
- Go to the `/site-profiler/:url` and enter a site URL
- Open devtools Network tab, and click the Check site button
- Check that the loading screen is shown
- Check that the `/basic` and `/advanced/performance` requests are sent out

https://github.com/Automattic/wp-calypso/assets/11555574/5cfc4e00-2fef-4f58-a55d-3391c3fb22e2

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

gcsecsey committed:
add loading screen component

gcsecsey committed:
show loading screen while performance metrics are fetching

gcsecsey committed:
animate progress bar while loading

gavande1 reviewed (approved):
LGTM 👍

epeicher commented:
Hi @gcsecsey, in Figma there is a comment that suggests switching between different sentences https://www.figma.com/design/P7Jjk4nbUTahUk8aHkG6NQ?node-id=1106-37002&m=dev#817920532. Do you plan to do it as part of a follow-up task?

gcsecsey commented:
> Hi @gcsecsey, in Figma there is a comment that suggests switching between different sentences https://www.figma.com/design/P7Jjk4nbUTahUk8aHkG6NQ?node-id=1106-37002&m=dev#817920532. Do you plan to do it as part of a follow-up task?

Sorry I didn't catch this before. I'll add it to this PR soon.

gcsecsey committed:
change headings with progress

epeicher commented:
Thanks for adding the other sentences! I have tested them and it seems there are some issues, I have tested with a random site, http://calypso.localhost:3000/site-profiler/https://optical.toys and this is the result after some seconds:

![CleanShot 2024-05-30 at 14 30 42](https://github.com/Automattic/wp-calypso/assets/3519124/0d68f2ed-b9c3-4f13-851f-0cc1e6be7ef4)


a

gcsecsey committed:
fix loading indicator

gcsecsey commented:
I've fixed the loading indicator not to produce any unexpected behavior. I also updated it to reset the progress when it's full, so that it always indicates that's something loading in the background.

I've found that the loading screen will be shown indefinitely for sites where we haven't yet queued the advanced tests. I took some site URLs from Site Profiles for testing:

https://github.com/Automattic/wp-calypso/assets/11555574/c604ff3d-4df2-4e4a-83c7-f67c0c7b7c7d

I'll update the branch to show the loading screen in the correct state for basic tests as well and push for another round of reviews.

gcsecsey committed:
update response type to match API response

gcsecsey commented:
I've updated the response types for the basic request to match the changes made in D150503-code.

When the CrUX data is not available for the site, the response will contain an `basic.data.errors` field and a `basic.data.success` field. This now need to be considered to avoid parsing errors.
<img width="1185" alt="image" src="https://github.com/Automattic/wp-calypso/assets/11555574/19296a57-ee57-4d59-9298-68c651d02b3d">

I updated the `showBasicMetrics` flag to also include the `basicMetrics.success` flag, when this isn't set, we couldn't parse the `BasicMetrics` data type.

cpapazoglou committed:
Site Profiler: progress bar styles.

epeicher committed:
Fix a typescript error

cpapazoglou committed:
Site Profiler: loading screen polling

epeicher reviewed (approved):
I have tested it after latest changes, and the polling works perfectly. After the last call with results `completed`, the results are displayed

![CleanShot 2024-05-31 at 18 48 16@2x](https://github.com/Automattic/wp-calypso/assets/3519124/61b242a7-0a2b-4806-8e35-5d44c8c1c760)
